src/strings/utf8.c

```diff
@@ -384,17 +384,18 @@ MVMuint32 MVM_string_utf8_decodestream(MVMThreadContext *tc, MVMDecodeStream *ds
                     break;
                 }
                 case UTF8_REJECT: {
-                    char *waste[] = { (char *)buffer, NULL };
-                    if (bufsize >= 3) {
-                        MVMGrapheme32 a = buffer[pos - 2], b = buffer[pos - 1], c = buffer[pos];
+                    MVMint32 error_pos = pos - 1;
+                    char *waste[] = { (char *)bytes, NULL };
+                    if (error_pos >= 3) {
+                        MVMGrapheme32 a = bytes[error_pos - 2], b = bytes[error_pos - 1], c = bytes[error_pos];
                         MVM_exception_throw_adhoc_free(tc, waste, "Malformed UTF-8 near bytes %02x %02x %02x", a, b, c);
```

Comment:
To avoid the mentioned sign extension, the format specifier should be %02hhx:
```
hh      Specifies that a following d, i, o, u, x, or X conversion specifier applies to a signed char or unsigned char argument  (the  argument
               will  have  been promoted according to the integer promotions, but its value shall be converted to signed char or unsigned char before
               printing);
```